experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts

```diff
@@ -0,0 +1,144 @@
+ * @alpha
+ */
+export function adaptEnum<TScope extends string, const TEnum extends Record<string, string>>(
+	factory: SchemaFactory<TScope>,
+	members: TEnum,
+) {
+	type Values = TEnum[keyof TEnum];
+	const values = Object.values(members) as Values[];
+	const inverse = new Map(Object.entries(members).map(([key, value]) => [value, key])) as Map<
+		Values,
+		keyof TEnum
+	>;
+
+	if (inverse.size !== values.length) {
+		throw new UsageError("All members of enums must have distinct values.");
+	}
+
+	type TOut = {
+		readonly [Property in keyof TEnum]: ReturnType<
+			typeof singletonSchema<TScope, TEnum[Property]>
+		>;
+	};
+	const factoryOut = <TValue extends Values>(value: TValue) => {
+		return new out[inverse.get(value) ?? fail("missing enum value")](
+			{},
+		) as unknown as ReturnType<typeof singletonSchema<TScope, TValue>>;
+	};
+	const out = factoryOut as typeof factoryOut & TOut;
+	for (const [key, value] of Object.entries(members)) {
+		Object.defineProperty(out, key, {
+			enumerable: true,
+			configurable: false,
+			writable: false,
+			value: singletonSchema(factory, value),
+		});
+	}
+
+	return out;
+}
+
+/**
+ * `Object.values`, but with more specific types.
+ * @alpha
+ */
+export function typedObjectValues<TKey extends string, TValues>(
+	object: Record<TKey, TValues>,
+): TValues[] {
+	return Object.values(object);
+}
+
+/**
+ * Converts an array of distinct strings into a collection of schema which can be used like an enum style union.
+ * @remarks
+ * The returned collection is also a function which can be used to convert strings into {@link Unhydrated} nodes in the union.
+ * Each node type has a `.value` getter which returns the associated string.
+ *
+ * The produced nodes use the provided strings as their `name`, and don't store any data beyond that.
+ * @alpha
+ */
+export function enumFromStrings<TScope extends string, const Members extends string>(
+	factory: SchemaFactory<TScope>,
+	members: Members[],
+) {
+	const names = new Set(members);
+	if (names.size !== members.length) {
+		throw new UsageError("All members of enums must have distinct names");
+	}
+
+	type TOut = Record<Members, ReturnType<typeof singletonSchema<TScope, Members>>>;
+	const factoryOut = <TValue extends Members>(value: TValue) => {
+		return new out[value]({}) as unknown as ReturnType<typeof singletonSchema<TScope, TValue>>;
+	};
+	const out = factoryOut as typeof factoryOut & TOut;
+	// const out: Record<Members, ReturnType<typeof makeSchema<Members>>> = Object.create(null);
+	for (const name of members) {
+		Object.defineProperty(out, name, {
+			enumerable: true,
+			configurable: false,
+			writable: false,
+			value: singletonSchema(factory, name),
+		});
+	}
+
+	return out;
+}
+
+// TODO: Why does this one generate an invalid d.ts file if exported?
+// Tracked by https://github.com/microsoft/TypeScript/issues/56718
+function _enumFromStrings2<TScope extends string, const Members extends readonly string[]>(
```

Comment:
Does this exist here _only_ to repro the above issue? Doesn't appear to be used within the module.

Reply:
This is a less code duplication heavy version of enumFromString. I've updated the TODO to better indicate this.